hs/src/Reach/Debugger.hs

```diff
@@ -105,315 +110,365 @@ type Id = Integer
+  -- (BYTES_CONCAT, [V_Bytes lhs,V_Bytes rhs]) -> return $ Succ $ V_Bytes $ (++) lhs rhs
 
 instance Interp DLArg where
   interp = \case
     DLA_Var dlvar -> do
       st <- asks e_store
-      return $ st M.! dlvar
+      return $ Succ $ st M.! dlvar
     -- TODO: fake constants per connector
-    DLA_Constant _dlconst -> return $ V_UInt $ toInteger (maxBound :: Int)
+    DLA_Constant _dlconst -> return $ Succ $ V_UInt $ toInteger (maxBound :: Int)
     DLA_Literal dllit -> interp dllit
     -- TODO: handle interact case: needs participant frontend state
     DLA_Interact _slpart _string _dltype -> undefined
 
 instance Interp DLLiteral where
   interp = \case
-    DLL_Null -> return $ V_Null
-    DLL_Bool bool -> return $ V_Bool bool
-    DLL_Int _at int -> return $ V_UInt int
-    DLL_Bytes bytes -> return $ V_Bytes $ show bytes
+    DLL_Null -> return $ Succ $ V_Null
+    DLL_Bool bool -> return $ Succ $ V_Bool bool
+    DLL_Int _at int -> return $ Succ $ V_UInt int
+    -- DLL_Bytes bytes -> return $ Succ $ V_Bytes $ show bytes
 
 instance Interp DLLargeArg where
   interp = \case
-    DLLA_Array _dltype dlargs -> V_Array <$> mapM interp dlargs
-    DLLA_Tuple dlargs -> V_Tuple <$> mapM interp dlargs
-    DLLA_Obj map_strs_to_dlargs -> V_Object <$> mapM interp map_strs_to_dlargs
+    DLLA_Array _dltype dlargs -> Succ <$> V_Array <$> map forceSucc <$> mapM interp dlargs
+    DLLA_Tuple dlargs -> Succ <$> V_Tuple <$> map forceSucc <$> mapM interp dlargs
+    DLLA_Obj strs_to_dlargs -> Succ <$> V_Object <$> M.map forceSucc <$> mapM interp strs_to_dlargs
     DLLA_Data _map_slvars_to_dltypes string dlarg -> do
       evd_arg <- interp $ dlarg
-      return $ V_Data string evd_arg
+      return $ Succ $ V_Data string $ forceSucc evd_arg
     DLLA_Struct assoc_slvars_dlargs -> do
       evd_args <- mapM (\arg -> interp arg) $ M.fromList assoc_slvars_dlargs
-      return $ V_Struct $ M.toList evd_args
+      return $ Succ $ V_Struct $ M.toList $ M.map forceSucc evd_args
 
 instance Interp DLExpr where
   interp = \case
     DLE_Arg _at dlarg -> interp dlarg
     DLE_LArg _at dllargearg -> interp dllargearg
-    DLE_Impossible at err -> expect_thrown at err
+    -- DLE_Impossible at err -> expect_thrown at err
     DLE_PrimOp _at primop dlargs -> do
-      evd_args <- mapM interp dlargs
+      evd_args <- map forceSucc <$> mapM interp dlargs
       interpPrim (primop,evd_args)
     DLE_ArrayRef _at dlarg1 dlarg2 -> do
-      ev1 <- interp dlarg1
-      ev2 <- interp dlarg2
+      ev1 <- forceSucc <$> interp dlarg1
+      ev2 <- forceSucc <$> interp dlarg2
       case (ev1,ev2) of
-        (V_Array arr, V_UInt n) -> return $ arr !! (fromIntegral n)
+        (V_Array arr, V_UInt n) -> return $ Succ $ arr !! (fromIntegral n)
         _ -> impossible "expression interpreter"
     DLE_ArraySet _at dlarg1 dlarg2 dlarg3 -> do
-      ev1 <- interp dlarg1
-      ev2 <- interp dlarg2
-      ev3 <- interp dlarg3
+      ev1 <- forceSucc <$> interp dlarg1
```

Comment:
Needing `forceSucc` is a sign that you haven't got the types of the continuation transformer cooperating enough. The whole idea of the monad is that it will hide the fact that capture and suspension happens beneath you. If you need to force it, then that means you're implementing things yourself

Reply:
(The same is true with needing `Succ` itself.) The `Result` would be something that is part of the state of a thread, not part of the interpreter

Reply:
@jeapostrophe oh i thought that was the design of the interpreter based on your code example: 

```

Jay
–
Oct 25th
View in channel
Jay 11:20 AM
interp :: a -> App Result

data Result
  = Succ DLVal
  | Block Signal (DLVal -> App Result)

``` 

so you're saying that the interpreter shouldn't return a `Result`?

Reply:
There's two options for the interpreter:
- "Hide blocking" --- Like by using a monad, so the interpreter will just return `DLVal` but the blocking will be hidden behind the scenes
- "Expose blocking" --- In this case, the `Result` type is useful because it allows us to do it by hand

Regardless of what we do for `interp`, there's another issue of how to keep track of what's going on in a single thread and it needs something like `Result` that says if it is done or if it is blocked. That is needed whether it is hidden or exposed in interp